sda/websocket.go

```diff
@@ -0,0 +1,366 @@
+		startstop: make(chan bool),
+	}
+	webHost, err := getWebHost(si)
+	log.ErrFatal(err)
+	w.mux = http.NewServeMux()
+	w.server = &graceful.Server{
+		Timeout: 100 * time.Millisecond,
+		Server: &http.Server{
+			Addr:    webHost,
+			Handler: w.mux,
+		},
+	}
+	return w
+}
+
+// Start listening on the port.
+func (w *WebSocket) Start() {
+	w.Lock()
+	w.started = true
+	w.Unlock()
+	log.Lvl3("Starting to listen on", w.server.Server.Addr)
+	go func() {
+		w.server.ListenAndServe()
+		w.startstop <- false
+	}()
+	w.startstop <- true
+}
+
+// RegisterService saves the service as being able to handle messages.
+func (w *WebSocket) RegisterService(service string, s Service) error {
+	w.services[service] = s
+	return nil
+}
+
+// RegisterMessageHandler for a service. Requests from a client to
+// "ws://service/path" will be forwarded to the corresponding
+// ProcessClientRequest of the 'service'.
+func (w *WebSocket) RegisterMessageHandler(service, path string) error {
+	log.Lvlf3("Registering websocket for ws://hostname/%s/%s", service, path)
+	h := &wsHandler{
+		ws:      w,
+		service: service,
+		path:    path,
+	}
+	w.mux.Handle(fmt.Sprintf("/%s/%s", service, path), h)
+	return nil
+}
+
+// Stop the websocket and free the port.
+func (w *WebSocket) Stop() {
+	w.Lock()
+	defer w.Unlock()
+	if !w.started {
+		return
+	}
+	log.Lvl3("Stopping", w.server.Server.Addr)
+	<-w.startstop
+	w.server.Stop(100 * time.Millisecond)
+	<-w.startstop
+	w.started = false
+}
+
+// Client is a struct used to communicate with a remote Service running on a
+// sda.Conode
+type Client struct {
+	service string
+	si      *network.ServerIdentity
+	conn    *websocket.Conn
+	// whether to keep the connection
+	keep bool
+	sync.Mutex
+}
+
+// NewClient returns a client using the service s. On the first Send, the
+// connection will be started, until Close is called.
+func NewClient(s string) *Client {
+	return &Client{
+		service: s,
+	}
+}
+
+// NewClientKeep returns a Client that doesn't close the connection between
+// two messages if it's the same conode.
+func NewClientKeep(s string) *Client {
+	return &Client{
+		service: s,
+		keep:    true,
+	}
+}
+
+// Send will marshal the message into a ClientRequest message and send it.
+func (c *Client) Send(dst *network.ServerIdentity, path string, buf []byte) ([]byte, ClientError) {
+	c.Lock()
+	defer c.Unlock()
+	if c.keep &&
+		(c.si != nil && !c.si.ID.Equal(dst.ID)) {
+		// We already have a connection, but to another ServerIdentity
+		c.Close()
+	}
+	if c.si == nil {
+		// Open connection to service.
+		url, err := getWebHost(dst)
+		if err != nil {
+			return nil, NewClientError(err)
+		}
+		log.Lvlf4("Sending %x to %s/%s/%s", buf, url, c.service, path)
+		d := &websocket.Dialer{}
+		// Re-try to connect in case the websocket is just about to start
+		for a := 0; a < 10; a++ {
+			c.conn, _, err = d.Dial(fmt.Sprintf("ws://%s/%s/%s", url, c.service, path),
+				http.Header{})
+			if err == nil {
+				break
```

Comment:
it's in the for-loop that finishes once a connection has been established or if too many trials finished.